net/libfab/libfab.c

```diff
@@ -0,0 +1,2963 @@
+ *
+ */
+static int libfab_ep_addr_decode(struct m0_fab__ep *ep, const char *name)
+{
+	char *node = ep->fep_name.fen_addr;
+	char *port = ep->fep_name.fen_port;
+	size_t nodeSize = ARRAY_SIZE(ep->fep_name.fen_addr);
+	size_t portSize = ARRAY_SIZE(ep->fep_name.fen_port);
+	int result;
+
+	M0_ENTRY("name=%s", name);
+	
+	if( name == NULL || name[0] == 0)
+		result =  M0_ERR(-EPROTO);
+	else if ((strncmp(name, "libfab", 6)) == 0)
+		result = libfab_ep_addr_decode_native(name, node, nodeSize, 
+						      port, portSize);
+	else if (name[0] < '0' || name[0] > '9')
+		/* sock format */
+		result = libfab_ep_addr_decode_sock(name, node, nodeSize, 
+						    port, portSize);
+	else
+		/* Lnet format. */
+		result = libfab_ep_addr_decode_lnet(name, node, nodeSize, 
+						    port, portSize);
+
+	if (result == FI_SUCCESS)
+		strcpy(ep->fep_name.fen_str_addr, name);
+
+	return M0_RC(result);
+}
+
+/**
+ * Used to lock the transfer machine mutex
+ */
+static inline void libfab_tm_lock(struct m0_fab__tm *tm)
+{
+	m0_mutex_lock(&tm->ftm_ntm->ntm_mutex);
+}
+
+/**
+ * Used to unlock the transfer machine mutex
+ */
+static inline void libfab_tm_unlock(struct m0_fab__tm *tm)
+{
+	m0_mutex_unlock(&tm->ftm_ntm->ntm_mutex);
+}
+
+static inline int libfab_tm_trylock(struct m0_fab__tm *tm)
+{
+	return m0_mutex_trylock(&tm->ftm_ntm->ntm_mutex);
+}
+
+/**
+ * Used to lock the transfer machine event post mutex
+ */
+static inline void libfab_tm_evpost_lock(struct m0_fab__tm *tm)
+{
+	m0_mutex_lock(&tm->ftm_evpost);
+}
+
+/**
+ * Used to unlock the transfer machine event post mutex
+ */
+static inline void libfab_tm_evpost_unlock(struct m0_fab__tm *tm)
+{
+	m0_mutex_unlock(&tm->ftm_evpost);
+}
+
+/**
+ * Used to check if the transfer machine mutex is locked.
+ * Returns true if locked, else false.
+ */
+static inline bool libfab_tm_is_locked(const struct m0_fab__tm *tm)
+{
+	return m0_mutex_is_locked(&tm->ftm_ntm->ntm_mutex);
+}
+
+/**
+ * Helper function that posts a tm state change event.
+ */
+static void libfab_tm_event_post(struct m0_fab__tm *tm,
+				 enum m0_net_tm_state state)
+{
+	struct m0_net_end_point *listen = NULL;
+
+	if (state == M0_NET_TM_STARTED) {
+		/* Check for LISTENING Passive endpoint */
+		listen = &tm->ftm_pep->fep_nep;
+		M0_ASSERT(listen != NULL);
+	}
+	
+	m0_net_tm_event_post(&(struct m0_net_tm_event) {
+		.nte_type       = M0_NET_TEV_STATE_CHANGE,
+		.nte_next_state = state,
+		.nte_time       = m0_time_now(),
+		.nte_ep         = listen,
+		.nte_tm         = tm->ftm_ntm,
+	});
+}
+
+/**
+ * Finds queued buffers that timed out and completes them with a
+ * ETIMEDOUT error.
+ */
+static void libfab_tm_buf_timeout(struct m0_fab__tm *ftm)
+{
+	struct m0_net_transfer_mc *net = ftm->ftm_ntm;
+	int                        i;
+	m0_time_t                  now = m0_time_now();
+
+	M0_PRE(libfab_tm_is_locked(ftm));
+	M0_PRE(libfab_tm_invariant(ftm));
+
+	for (i = 0; i < ARRAY_SIZE(net->ntm_q); ++i) {
```

Comment:
This can be time consuming function as it goes through all the buffers in these 6 queue (	M0_NET_QT_MSG_RECV, 	M0_NET_QT_MSG_SEND, M0_NET_QT_PASSIVE_BULK_RECV, M0_NET_QT_PASSIVE_BULK_SEND, M0_NET_QT_ACTIVE_BULK_RECV, M0_NET_QT_ACTIVE_BULK_SEND).

We can stagger the timeout checks e.g. not all queue need to be checked once, e.g. every 100/6 milisecond check one queue and then move to next elapsed interval of 100/6, check next queue and so on.
- Goal is to avoid checking all the queue at once
- Avoid checking queue in every poll run

Reply:
As discussed, added code to check buffer timeouts after 1 sec.